File: 02-tech/01-programming-fundamentals/13-exam/03-softuni-bar-income/SoftuniBarIncome.cs
using System;
using System.Text.RegularExpressions;

namespace _03_softuni_bar_income
{
    class SoftuniBarIncome
    {
        // %([A-Z][a-z]+)%.*<(\w+)>.*\|(\d+)\|\D*(\d+(\.\d+)?)\$
        static void Main()
        {
            var namePattern = @"%([A-Z][a-z]+)%";
            var productPattern = @"<(\w+)>";
            var countPattern = @"\|(\d+)\|";
            var pricePattern = @"(\d+(\.\d+)?)\$";
            var validOrderRegex = new Regex($@"{namePattern}.*{productPattern}.*{countPattern}\D*{pricePattern}");
            var totalIncome = 0M;

            string input = null;
            while ((input = Console.ReadLine()) != "end of shift")
            {
                var orderMatch = validOrderRegex.Match(input);
                if (orderMatch.Success)
                {
                    var customer = orderMatch.Groups[1].Value;
                    var product = orderMatch.Groups[2].Value;
                    var count = int.Parse(orderMatch.Groups[3].Value);
                    var price = decimal.Parse(orderMatch.Groups[4].Value);

                    var bill = price * count;
                    Console.WriteLine($"{customer}: {product} - {bill:0.00}");

                    totalIncome += bill;
                }
            }

            Console.WriteLine($"Total income: {totalIncome:0.00}");
        }
    }
}
/*

%George%<Croissant>|2|10.3$
%Peter%<Gum>|1|1.3$
%Maria%<Cola>|1|2.4$
end of shift

%InvalidName%<Croissant>|2|10.3$
%Peter%<Gum>1.3$
%Maria%<Cola>|1|2.4
%Valid%<Valid>valid|10|valid20$
end of shift

 */

File: 02-tech/01-programming-fundamentals/13-exam/03-softuni-bar-income/softuni-bar-income.js
// jshint esversion:6
function solve(orders)
{
    // %([A-Z][a-z]+)%.*<(\w+)>.*\|(\d+)\|\D*(\d+(\.\d+)?)\$
    orders.pop();
    let income = 0;
    var namePattern = '%([A-Z][a-z]+)%';
    var productPattern = '<(\\w+)>';
    var countPattern = '\\|(\\d+)\\|';
    var pricePattern = '(\\d+(\\.\\d+)?)\\$';
    var validOrderRegex = new RegExp(`${namePattern}.*${productPattern}.*${countPattern}\\D*${pricePattern}`);

    for (const order of orders)
    {
        const match = validOrderRegex.exec(order);
        if (match)
        {
            const [_, customer, product, quantity, price] = match;
            const bill = (+price) * (+quantity);

            console.log(`${customer}: ${product} - ${bill.toFixed(2)}`);

            income += bill;
        }
    }

    console.log(`Total income: ${income.toFixed(2)}`);
}

// solve([
//     '%George%<Croissant>|2|10.3$',
//     '%Peter%<Gum>|1|1.3$',
//     '%Maria%<Cola>|1|2.4$',
//     'end of shift'
// ]);

// solve([
//     '%InvalidName%<Croissant>|2|10.3$',
//     '%Peter%<Gum>1.3$',
//     '%Maria%<Cola>|1|2.4',
//     '%Valid%<Valid>valid|10|valid20$',
//     'end of shift'
// ]);